### PR TITLE
sdjournal: add formating feature to JournalReader

### LINKED ISSUE #253: Feature request: Custom formatter for sdjournal reader
At the moment, `JournalReader` has a fixed message template for each entry. It would be really helpful if I would be able to alter the format.

I have thought of something like this:

```golang
config := sdjournal.JournalReaderConfig {
		Formatter: func(entry *JournalEntry) string {
                     return fmt.Sprintf("%s:%s",entry.Fields["MY_CUSTOM_FIELD"])
                },
	}
```

I can send you a patch if you are okay with this kind of feature.

### PR DESCRIPTION
**Pull request by Richterrettich**
this allows users of JournalReader to provide a callback function
that will format the output of each entry.

Fixes #253

### CODE REVIEW

File: sdjournal/read.go [hunk 1]
```diff
@@ -44,19 +44,27 @@ type JournalReaderConfig struct {
 	// If not empty, the journal instance will point to a journal residing
 	// in this directory. The supplied path may be relative or absolute.
 	Path string
+
+	// If not nil, Formatter will be used to translate the resulting entries
+	// into strings. If not set, the default format (timestamp and message field)
+	// will be used.
+	Formatter func(entry *JournalEntry) string
```
Comment (lucab) [inline sdjournal/read.go]: I would prefer to make this more error-friendly by returning a `(string, error)` instead. As a consumer, would it be ok?

File: sdjournal/read.go [hunk 2]
```diff
@@ -245,6 +253,13 @@ func (r *JournalReader) buildMessage() (string, error) {
 	var usec uint64
 	var err error
 
+	if r.formatter != nil {
+		entry, err := r.journal.GetEntry()
```
Comment (lucab) [inline sdjournal/read.go]: Given that we are introducing a formatting function, I would suggest to lift the entry fetching logic into the common path at `Read`, then change the `buildMessage` signature to match the formatter one, and just call `r.formatter(entry)` there.

**Comment by lucab** [comment]
Thanks for your PR! I generally like the idea, I've just left a couple of comments to streamline the public interface and the integration point.

**Comment by Richterrettich** [comment]
Thanks for the review. I will rework the patch and inlcude your suggestions.

**Comment by Richterrettich** [comment]
ping @lucab , I think I have addressed all your suggestions. Feel free to take a look.

File: sdjournal/read.go [hunk 3]
```diff
@@ -44,19 +44,33 @@ type JournalReaderConfig struct {
...
 	journal   *Journal
 	msgReader *strings.Reader
+	formatter func(entry *JournalEntry) (string, error)
 }
 
 // NewJournalReader creates a new JournalReader with configuration options that are similar to the
 // systemd journalctl tool's iteration and filtering features.
 func NewJournalReader(config JournalReaderConfig) (*JournalReader, error) {
-	r := &JournalReader{}
+
```
Comment (lucab) [inline sdjournal/read.go]: Spurious leading newline.

File: sdjournal/read.go [hunk 4]
```diff
@@ -238,20 +257,12 @@ process:
...
-
-	if msg, err = r.journal.GetData("MESSAGE"); err != nil {
-		return "", err
-	}
-
-	if usec, err = r.journal.GetRealtimeUsec(); err != nil {
-		return "", err
-	}
+func simpelMessageFormatter(entry *JournalEntry) (string, error) {
+	msg := entry.Fields["MESSAGE"]
```
Comment (lucab) [inline sdjournal/read.go]: Please check for key presence:
```golang
msg, ok := entry.Fields["MESSAGE"]
if !ok { ... }
```

**Comment by Richterrettich** [comment]
Yeah no problem. I somehow thougth that the "MESSAGE" field could never be empty in a journal entry. But a check can't hurt. I fixed both issues. Can you take a look again please?

**Pull request merged**